core/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala

```diff
@@ -140,12 +140,16 @@ case class WriteIntoDelta(
       val useDynamic = txn.metadata.partitionColumns.nonEmpty &&
         options.isDynamicPartitionOverwriteMode
       options.replaceWhere.foreach { _ =>
-        if (useDynamic) {
-          throw new AnalysisException(s"'${DeltaOptions.REPLACE_WHERE_OPTION}' cannot be used" +
-            s" when '${DeltaOptions.PARTITION_OVERWRITE_MODE_OPTION}' is set to 'DYNAMIC'")
+        if (useDynamic && options.partitionOverwriteModeInOptions) {
```

Comment:
this logic is a bit hard to follow as there is implicit assumption here that if partition overwrite mode is set in an option AND dynamic is effectively enabled (as a result of the complex code in DeltaOptions).. that means that dynamic must be enabled via the option. this is not obvious from the code.

i would write it such that its more obvious... like
```
if (isTablePartitioned && isDynamicPartitionOverwriteSetInOption && isReplaceWhereSetInOption)
```

its more verbose to create these fields .. but the results is more readable and obvious.